### PR TITLE
GH-20415: [Go] Kernel Input Type for RLE

### PR DESCRIPTION
**Pull request by zeroshade**
* Closes: #20415

### CODE REVIEW

**Comment by zeroshade** [comment]
CC @zagto

File: go/arrow/array/concat.go [hunk 1]
```diff
@@ -42,7 +43,12 @@ func Concatenate(arrs []arrow.Array, mem memory.Allocator) (result arrow.Array,
 
 	defer func() {
 		if pErr := recover(); pErr != nil {
-			err = fmt.Errorf("arrow/concat: unknown error: %v", pErr)
+			switch e := pErr.(type) {
+			case error:
```
Comment (westonpace) [inline go/arrow/array/concat.go:47]: I don't really know go but why is it important to distinguish between these two error types?  Adding `unknown error` doesn't seem to contribute too much to a user's understanding of the error.

File: go/arrow/array/concat.go [hunk 2]
```diff
@@ -512,6 +518,78 @@ func concat(data []arrow.ArrayData, mem memory.Allocator) (arrow.ArrayData, erro
...
+			off := rle.GetPhysicalOffset(d)
+
+			runs[i] = NewSliceData(d.Children()[0], int64(off), int64(off+plen))
+			defer runs[i].Release()
+			values[i] = NewSliceData(d.Children()[1], int64(off), int64(off+plen))
+			defer values[i].Release()
+
+			physicalLength, overflow = addOvf32(physicalLength, int32(plen))
+			if overflow {
+				return nil, fmt.Errorf("%w: run length encoded array length must fit into a 32-bit signed integer",
```
Comment (westonpace) [inline go/arrow/array/concat.go:538]: Does go use 64-bit lengths for arrays?

File: go/arrow/array/concat_test.go [hunk 3]
```diff
@@ -568,3 +569,59 @@ func TestConcatDictionaryNullSlots(t *testing.T) {
 
 	assert.Truef(t, array.Equal(actual, expected), "got: %s, expected: %s", actual, expected)
 }
+
+func TestConcatRunLengthEncoded(t *testing.T) {
```
Comment (westonpace) [inline go/arrow/array/concat_test.go:573]: Do you want any tests for the overflow cases?

File: go/arrow/array/encoded.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+func (b *RunLengthEncodedBuilder) Reserve(n int) {
+	b.values.Reserve(n)
+	b.runEnds.Reserve(n)
+}
+
+func (b *RunLengthEncodedBuilder) Resize(n int) {
+	b.values.Resize(n)
+	b.runEnds.Resize(n)
+}
```
Comment (westonpace) [inline go/arrow/array/encoded.go:248]: Do you not need to resize/reserve the validity map?

File: go/arrow/array/array.go [hunk 5]
```diff
@@ -175,8 +175,8 @@ func init() {
 		arrow.LARGE_STRING:            func(data arrow.ArrayData) arrow.Array { return NewLargeStringData(data) },
 		arrow.LARGE_BINARY:            func(data arrow.ArrayData) arrow.Array { return NewLargeBinaryData(data) },
 		arrow.LARGE_LIST:              func(data arrow.ArrayData) arrow.Array { return NewLargeListData(data) },
-		arrow.INTERVAL:                func(data arrow.ArrayData) arrow.Array { return NewIntervalData(data) },
```
Comment (westonpace) [inline go/arrow/array/array.go:178]: I assume getting rid of interval is an intentional unrelated change and not a mistake?

File: go/arrow/rle/rle_utils_test.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		{[]int32{1, 2, 3}, 2, 2},
+		{[]int32{2, 3, 4}, 0, 0},
+		{[]int32{2, 3, 4}, 1, 0},
+		{[]int32{2, 3, 4}, 2, 1},
+		{[]int32{2, 3, 4}, 3, 2},
+		{[]int32{2, 4, 6}, 3, 1},
+		{[]int32{1, 2, 3, 4, 5, 6, 7, 8, 9, 1000, 1005, 1015, 1020, 1025, 1050}, 1000, 10},
+		// out-of-range logical offset should return len(vals)
+		{[]int32{2, 4, 6}, 6, 3},
+		{[]int32{2, 4, 6}, 10000, 3},
```
Comment (westonpace) [inline go/arrow/rle/rle_utils_test.go:49]: What about out-of-range in the negative direction?  Error?

**Review by westonpace — COMMENTED** [review]
Looks fairly straightforward, though I don't truly understand go

File: go/arrow/array/concat.go (hunk 2 above)
Comment (zeroshade) [reply go/arrow/array/concat.go:538]: The Go impl uses `int` which will be 32-bit on 32-bit systems and 64-bit on 64-bit systems.

File: go/arrow/array/concat_test.go (hunk 3 above)
Comment (zeroshade) [reply go/arrow/array/concat_test.go:573]: That's a good idea, i'll add a test for that

File: go/arrow/array/encoded.go (hunk 4 above)
Comment (zeroshade) [reply go/arrow/array/encoded.go:248]: based on the current spec `RunLengthEncoded` arrays have no validity map

File: go/arrow/array/array.go (hunk 5 above)
Comment (zeroshade) [reply go/arrow/array/array.go:178]: yes. If you look at `datatype.go` you can see there's a comment there pointing out that `INTERVAL` as an enum value is deprecated in favor of the individual interval enums (`INTERVAL_MONTH_DAY_NANO`, etc...) So adding the new `RUN_LENGTH_ENCODED` type seemed like a good time to clean that out.

File: go/arrow/rle/rle_utils_test.go (hunk 6 above)
Comment (zeroshade) [reply go/arrow/rle/rle_utils_test.go:49]: I'll add a test case for it, I originally just copied these tests over from the C++ draft PR :smile:

File: go/arrow/array/concat.go (hunk 1 above)
Comment (zeroshade) [reply go/arrow/array/concat.go:47]: using `%w` wraps the error that was panic'd so that someone receiving the error can use `errors.Is` or `errors.As` if they desire to react differently based on the type of the error (such as `arrow.ErrInvalid` vs `arrow.ErrIndex` vs `arrow.ErrNotImplemented` etc...) basically it creates an error chain

In the case of the `"unknown error"` it means that we had a panic with a string rather than a proper error instance. Though I agree that the text "unknown error" doesn't contribute much to a user's understanding here, so i'll just get rid of that.

File: go/arrow/rle/rle_utils_test.go (hunk 6 above)
Comment (zeroshade) [reply go/arrow/rle/rle_utils_test.go:49]: Hmm, what do we think *should* happen? going out of range in the positive direction just returns `len(vals)` technically the offset should never be negative, rather than error I think the better solution here would be to just make the argument a `uint` rather than an `int`, thoughts?

File: go/arrow/compute/internal/exec/kernel.go [hunk 7]
```diff
@@ -242,6 +242,40 @@ func (primitiveMatcher) Equals(other TypeMatcher) bool {
...
+func (r reeMatcher) Equals(other TypeMatcher) bool {
+	o, ok := other.(reeMatcher)
+	if !ok {
+		return false
+	}
+	return r.runEndsMatcher.Equals(o.runEndsMatcher) && r.encodedMatcher.Equals(o.encodedMatcher)
+}
+
+func (r reeMatcher) String() string {
+	return "run_length_encoded(" + r.encodedMatcher.String() + ")"
```
Comment (lidavidm) [inline go/arrow/compute/internal/exec/kernel.go:268]: should runEndsMatcher be included in the repr?

**Review by lidavidm — APPROVED** [review]

File: go/arrow/compute/internal/exec/kernel.go (hunk 7 above)
Comment (zeroshade) [reply go/arrow/compute/internal/exec/kernel.go:268]: good point, it should. i'll fix that

**Comment by zeroshade** [comment]
I'll merge this at EOD if no one has any objections/comments
